### PR TITLE
[16.0][MIG] product_cost_security: Migration to 16.0

### PR DESCRIPTION
**Pull request by bvanjeelharia**

### CODE REVIEW

File: product_cost_security/views/product_views.xml [hunk 1]
```diff
@@ -6,21 +6,23 @@
...
             </xpath>
             <xpath expr="//div[@name='standard_price_uom']" position="attributes">
                 <attribute
-                    name="groups"
-                    separator=","
-                    add="product_cost_security.group_product_cost"
-                />
+                    name="attrs"
+                >{'invisible': [('user_can_see_standard_price_uom', '=', False),'|',('product_variant_count', '&gt;',
+                    1), ('is_product_variant', '=', False)]}
```
Comment (astirpe) [inline product_cost_security/views/product_views.xml]: This condition could be moved into `_compute_user_can_see_standard_price_uom()`

File: product_cost_security/tests/test_product_cost_security.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            # as the user does not have the group- Modify product costs
+            sheet_form.standard_price = 5.0
+        self.env.user.groups_id = [
+            (
+                6,
+                0,
+                [self.product_edit_cost_group_id],
+            )
+        ]
+        sheet_form = Form(self.product_template_model.with_user(self.env.user))
```
Comment (astirpe) [inline product_cost_security/tests/test_product_cost_security.py:73]: You are testing three different cases here, so you could split this test and have 3 tests in total

**Review by astirpe — CHANGES_REQUESTED** [review]

**Comment by astirpe** [comment]
You squashed the V16 changes into the pre-commit

File: product_cost_security/tests/test_product_cost_security.py (hunk 2 above)
Comment (bvanjeelharia) [reply product_cost_security/tests/test_product_cost_security.py:73]: Hello @astirpe , I have made the changes as per your suggestions.There are now 7 tests in all, as I had to add a field 'user_can_see_standard_price_uom' for product.product (product variant) as well.

Thanks

File: product_cost_security/models/product_template.py [hunk 3]
```diff
@@ -17,3 +20,11 @@ def _compute_user_can_update_cost(self):
...
                 "product_cost_security.group_product_edit_cost"
             )
+
+    @api.depends_context("uid")
+    def _compute_user_can_see_standard_price_uom(self):
+        """A user can see cost only when it has required permissions."""
+        for product in self:
+            product.user_can_see_standard_price_uom = self.env.user.has_group(
+                "product_cost_security.group_product_cost"
+            ) and (product.product_variant_count <= 1 or product.is_product_variant)
```
Comment (astirpe) [inline product_cost_security/models/product_template.py]: This method should depend on `"product_variant_count"` and `"is_product_variant"`

File: product_cost_security/models/product_product.py [hunk 4]
```diff
@@ -1,9 +1,20 @@
...
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
-from odoo import fields, models
+from odoo import api, fields, models
 
 
 class ProductProduct(models.Model):
     _inherit = "product.product"
 
     standard_price = fields.Float(groups="product_cost_security.group_product_cost")
+    user_can_see_standard_price_uom = fields.Boolean(
```
Comment (astirpe) [inline product_cost_security/models/product_product.py]: Why did you add this field? Is not used in the product.product view

**Review by astirpe — CHANGES_REQUESTED** [review]

File: product_cost_security/models/product_template.py (hunk 3 above)
Comment (bvanjeelharia) [reply product_cost_security/models/product_template.py]: Yes, I will make that change

File: product_cost_security/models/product_product.py (hunk 4 above)
Comment (bvanjeelharia) [reply product_cost_security/models/product_product.py]: Its because the view is common for both product.product and product.template

Comment (bvanjeelharia) [reply product_cost_security/models/product_product.py]: and the field was getting calculated based on product.template instead of product.product

File: product_cost_security/views/product_views.xml [hunk 5]
```diff
@@ -6,21 +6,22 @@
         <field name="arch" type="xml">
             <xpath expr="//div[@name='standard_price_uom']" position="inside">
                 <field name="user_can_update_cost" invisible="1" />
+                <field name="user_can_see_standard_price_uom" invisible="1" />
             </xpath>
             <xpath expr="//div[@name='standard_price_uom']" position="attributes">
                 <attribute
-                    name="groups"
-                    separator=","
-                    add="product_cost_security.group_product_cost"
```
Comment (astirpe) [inline product_cost_security/views/product_views.xml]: In the V15 version, group `product_cost_security.group_product_cost` was added to already existing "`base.group_user`"  for this field. So in theory you may have multiple groups for `standard_price_uom`.
But in `_compute_user_can_see_standard_price_uom()` you only handle one group. Why did you add `product_variant_count` and `is_product_variant` in the condition? Can't you just leave things as they were in V15?

**Review by astirpe — CHANGES_REQUESTED** [review]

File: product_cost_security/views/product_views.xml (hunk 5 above)
Comment (bvanjeelharia) [reply product_cost_security/views/product_views.xml]: Sorry for the inconveniences caused, this happened as the Odoo base addons were recently changed(https://github.com/odoo/odoo/commit/206e88b0ad9fdb6e1767c93b3eca8feeecddd3da) and I had an old db where the group was still applied on the field and it was causing an additional field (uom) to show up, please check attached screenshot where in the view is disturbed with per Units showing up which should have hidden. I have restored it to how it was in v15
Screenshot from 2023-01-03 16-50-57

**Review by astirpe — APPROVED** [review]

**Comment by bvanjeelharia** [comment]
Can someone please review and approve this PR?

**Comment by rafaelbn** [comment]
Please @bvanjeelharia , could you make a "commit" empty to try un block runboat in order to test it?

@sergio-teruel you could be interested in this PR 😄

**Comment by sergio-teruel** [comment]
@rafaelbn please give me the task :smile:

File: product_cost_security/tests/test_product_cost_security.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# Copyright 2023 Onestein (<https://www.onestein.eu>)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo.tests.common import Form, TransactionCase
+
+
+class TestProductCostSecurity(TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
Comment (sergio-teruel) [inline product_cost_security/tests/test_product_cost_security.py:10]: You can create an environment with tracking disable.
cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))

File: product_cost_security/tests/test_product_cost_security.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        sheet_form = Form(self.product_template_model.with_user(self.env.user))
+        sheet_form.name = "Test Product"
+        sheet_form.save()
+        with self.assertRaises(AssertionError):
+            # It would raise an AssertionError as the field would not be found in the view
+            # as the user does not have the group- Access to product costs
+            standard_price = sheet_form.standard_price
+            self.assertEqual(standard_price, 0.0)
+
+    def test_1(self):
```
Comment (sergio-teruel) [inline product_cost_security/tests/test_product_cost_security.py]: Can you use more descriptive names?

**Review by sergio-teruel — APPROVED** [review]
Minor changes
Please, can you modify the commit deleting the extra lines?

File: product_cost_security/tests/test_product_cost_security.py (hunk 6 above)
Comment (bvanjeelharia) [reply product_cost_security/tests/test_product_cost_security.py:10]: Made the change

File: product_cost_security/tests/test_product_cost_security.py (hunk 7 above)
Comment (bvanjeelharia) [reply product_cost_security/tests/test_product_cost_security.py]: Yes, I have made the changes

**Review by yajo — APPROVED** [review]

**Comment by CasVissers-360ERP** [comment]
@rafaelbn can be merged?

**Review by rafaelbn — APPROVED** [review]
Thank you! 👍

**Comment by bvanjeelharia** [comment]
Can this be merged now?

**Review by Yadier-Tecnativa — COMMENTED** [review]
check
